cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/type/DatasetTypeManager.java

```diff
@@ -372,7 +373,7 @@ public Boolean call() throws DatasetModuleConflictException, IOException {
               impersonator.doAs(datasetModuleId.getParent(), new Callable<Location>() {
                 @Override
                 public Location call() throws Exception {
-                  return locationFactory.create(module.getJarLocation());
+                  return Locations.getLocationFromAbsolutePath(locationFactory, module.getJarLocationPath());
```

Comment:
users of the DatasetModuleMeta class should not have to know they have to use this utility method. You should put a `getLocation(LocationFactory)` method in DatasetModuleMeta that handles the logic, then these classes call that method.